### PR TITLE
LazyInterpret for FeedVariableOpExpr

### PR DESCRIPTION
**Pull request by chengtbf**
LazyInterpret 支持 传入 EagerTensor 构造 VariableOp

### CODE REVIEW

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp [hunk 1]
```diff
@@ -113,9 +113,63 @@ Maybe<void> LazyInterpreter::ApplyImpl(const FeedInputOpExpr& op_expr, const Ten
...
+  const auto& parallel_attr =
+      JUST(compatible_py::GetOpArgParallelAttribute(blob_parallel_desc_sym, op_attr, obn));
+  const auto& blob_attr = JUST(compatible_py::GetOpArgBlobAttribute(op_attr, obn));
+
+  CHECK_OR_RETURN(!outputs->at(0).get());
+  (*outputs)[0] = JUST(OpInterpUtil::BuildTensor(blob_attr, parallel_attr, /*is_lazy=*/true));
+  // NOTE(chengcheng): Record variable op output LazyTenosr
+  TensorNameScope::Global()->Record(outputs->at(0), op_name + "/" + obn);
+  // NOTE(chengcheng): Record EagerTensor as variable tensor name
+  TensorNameScope::Global()->Record(input_tensor, op_name + "/" + obn);
```
Comment (chengtbf) [inline oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:172]: 这里我会把输入的 EagerTensor 也记录下来，这样后续的 UserOp LazyInterpret 里的 input 如果是 EagerTensor 也能找到正确的 lbn。

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp [hunk 2]
```diff
@@ -113,9 +113,63 @@ Maybe<void> LazyInterpreter::ApplyImpl(const FeedInputOpExpr& op_expr, const Ten
...
+  input_tensor->shape()->ToProto(var_conf->mutable_shape());
+  var_conf->set_data_type(input_tensor->dtype());
+  // NOTE(chengcheng): VariableOpConf initializer_conf is useless because variable is inited
+  //   by EagerTensor.
+  var_conf->mutable_initializer()->mutable_empty_conf();
+  if (input_tensor->is_consistent()) {
+    // TODO(chengcheng): GenerateParallelDistributionString by tensor.
+  }
+  if (!input_tensor->requires_grad()) { var_conf->set_trainable(false); }
+  // TODO(chengcheng, xuxiaoyu): Set L1/L2 RegularizerConf by nn.Graph Optimizer
```
Comment (chengtbf) [inline oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: 这里我列了一个 TODO： 现在单凭 EagerTensor 是不知道 Variable 的 L1 和 L2 参数的，PyTorch 的应该是放在了 Optimizer 里？ @strint  啸宇你后续研究一下，看 nn.Graph 如何支持配置每个 Variable 的 L1 和 L2

Comment (strint) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: 是不同参数配置不同的learning rate么

可以不同参数被不同optimizer绑定，每个opt一个lr
https://pytorch.org/docs/stable/optim.html

Comment (strint) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: 还有这种：
```
optim.SGD([
                {'params': model.base.parameters()},
                {'params': model.classifier.parameters(), 'lr': 1e-3}
            ], lr=1e-2, momentum=0.9)
```
optimizer中把参数分为多个group，每个group一个lr，而且还有个默认的lr

Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: 不是，LR 是 learning rate，我说的这个是 l1、l2 正则化的参数

Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: https://tensorflow.google.cn/api_docs/python/tf/keras/regularizers/L1L2?hl=en

Comment (strint) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: torch 没有统一，都可以在定义loss时，自己手动写，另外标准的写法：
- 对于l2，写在optimizer的weight_decay中；
- 对于l1，写在loss中；
参考：https://stackoverflow.com/a/66543549/2038618

Comment (strint) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: 看我们是否有必要，在optimizer的分group的参数配置中，加一个 l1_norm/l2_norm参数，进行配置 @wyg1997

Comment (wyg1997) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:143]: 应该是所有Optimizer构造时传的参数，都支持单独指定，如果l1、l2在Optimizer的参数列表里，每个ParamGroup都也应该支持的

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp (hunk 1 above)
Comment (leaves-zwx) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:172]: variable 的 lazy tensor 和 eager tensor 的内存共享会在哪里实现？

File: oneflow/python/test/graph/test_variable_op_expr.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            attrs = oneflow._oneflow_internal.MutableCfgAttrMap()
+
+            if not x.is_determined:
+                x.determine()
+            x_tensor_in_c = x._local_or_consistent_tensor
+
+            out_tensor = var_op.apply([x_tensor_in_c], attrs)[0]
+            test_case.assertEqual(out_tensor.shape, (1, 1, 10, 10))
+            test_case.assertTrue(out_tensor.is_lazy)
+            test_case.assertTrue(out_tensor.is_consistent)
```
Comment (strint) [inline oneflow/python/test/graph/test_variable_op_expr.py:85]: 当前默认产出都是consistent的？

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp [hunk 4]
```diff
@@ -113,9 +113,63 @@ Maybe<void> LazyInterpreter::ApplyImpl(const FeedInputOpExpr& op_expr, const Ten
...
+  CHECK_EQ_OR_RETURN(op_expr.input_size(), 1);
+  const std::shared_ptr<Tensor>& input_tensor = inputs.at(0);
+  CHECK_OR_RETURN(input_tensor->is_eager());
+
+  const auto& scope = JUST(GetCurrentScope());
+  int64_t scope_symbol_id = JUST(scope->symbol_id());
+
+  OperatorConf op_conf;
+  op_conf.set_name(op_expr.op_name());           // construct by python nn.Graph
+  op_conf.set_scope_symbol_id(scope_symbol_id);  // TODO(chengcheng): NewScope by cur scope.
```
Comment (strint) [inline oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:127]: 外部创建了一个对该variable的scope，因为是单Tensor的，貌似可以复用

**Review by strint — APPROVED** [review]

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp (hunk 1 above)
Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:172]: NNGraph 里，LazyInterpret 不关心这个事情。  由 python 端的 nn.Graph 把 Variable 的 EagerTensor 和 对应的 names 传进 NNGraph 里，NNGraph记录该信息，在 Runtime 启动时绑定 Regst。

File: oneflow/python/test/graph/test_variable_op_expr.py (hunk 3 above)
Comment (chengtbf) [reply oneflow/python/test/graph/test_variable_op_expr.py:85]: Lazy 其实只有 Consistent 概念； Mirror 也是展成 Consistent 的，即使你传进去的 是 一个 local tensor，也会翻译成 placement 是本 rank 的 ConsistentTensor

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp (hunk 4 above)
Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:127]: 需要，外部你创建的 scope 是 Block 的，没有真实的 ParallelDesc 信息。ParallelDesc 相关的 Scope 一定需要在 LazyInterpret· 里现场创建，无论是 Input、Variable 还是普通的 UserOp，因为这些都在输入的 Tensor 上保存的。

Comment (strint) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:127]: 主要原因是我们现在要根据tensor去推理ParallelDesc，所以Block层面创建的Scope里面的ParallelDesc往往就没用了对吧

Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:127]: 是的

**Review by leaves-zwx — APPROVED** [review]

File: oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp [hunk 5]
```diff
@@ -113,9 +113,63 @@ Maybe<void> LazyInterpreter::ApplyImpl(const FeedInputOpExpr& op_expr, const Ten
...
+  // Check outputs num and setup output tensor properties.
+  CHECK_EQ_OR_RETURN(outputs->size(), 1);
+  CHECK_EQ_OR_RETURN(op_expr.output_size(), 1);
+
+  const std::string obn = "out";  // NOTE(chengcheng): obn is NOT op_expr.indexed_obns
+  const auto& parallel_attr =
+      JUST(compatible_py::GetOpArgParallelAttribute(blob_parallel_desc_sym, op_attr, obn));
+  const auto& blob_attr = JUST(compatible_py::GetOpArgBlobAttribute(op_attr, obn));
+
+  CHECK_OR_RETURN(!outputs->at(0).get());
```
Comment (hjchen2) [inline oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:167]: 这里为什么要check outputs[0]

Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:167]: 你是说如果 outputs 有，就重置其 Shape？

Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:167]: 因为你下面留了一个 TODO，我这里就假设了要创建一个新的 Tensor

Comment (chengtbf) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:167]: 参考了你在这个文件里的 223 行

Comment (hjchen2) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:167]: 哦哦，看错了，我以为你这里需要外面传一个非空tensor进来

Comment (hjchen2) [reply oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:167]: >>> 你是说如果 outputs 有，就重置其 Shape？

不不，你现在这样就可以，我本来想的也是这样。outputs是非空的tensor表明是inplace的逻辑，FeedVar显然不会有in p la ce的。
